File: tools/gate/gate.ts
import { readFileSync } from "node:fs";
import { join } from "node:path";
import process from "node:process";
import { loadConfig } from "./config.ts";
import { sha256Normalized } from "./hash.ts";
import { isAtLeast, MIN_NODE } from "./node-version.ts";
import { recordsDir, repoRootFromGateFile } from "./paths.ts";

function writeOut(text: string): void {
  process.stdout.write(text);
}

function writeErr(text: string): void {
  process.stderr.write(text);
}

function requireNodeVersion(): number {
  const current = process.versions.node;
  if (isAtLeast(current, MIN_NODE)) return 0;
  writeErr(
    `keel: Node ${current} is below ${MIN_NODE} (DEC-150). Refusing to run.\n`,
  );
  return 1;
}

function cmdStatus(root: string): number {
  const records = recordsDir(root);
  const cfg = loadConfig(records);
  const tier =
    typeof cfg.enforcement_tier === "string" ? cfg.enforcement_tier : "unknown";
  writeOut("keel status\n");
  writeOut("wave: W1-stub\n");
  writeOut(`runtime: node+ts ${process.versions.node}\n`);
  writeOut(`root: ${root}\n`);
  writeOut(`records_dir: ${records}\n`);
  writeOut(`handoff: ${join(records, "handoff.md")}\n`);
  writeOut(`overview: ${join(records, "OVERVIEW.md")}\n`);
  writeOut(`plan_index: ${join(records, "plan", "INDEX.md")}\n`);
  writeOut(`enforcement_tier: ${tier}\n`);
  writeOut(
    "note: full gates are W2; this stub prints start-of-session paths (C-27/C-72).\n",
  );
  return 0;
}

function cmdHash(args: string[]): number {
  if (args.length !== 1) {
    writeErr("usage: gate hash <file>\n");
    return 2;
  }
  const target = args[0] ?? "";
  const bytes = readFileSync(target);
  writeOut(`${sha256Normalized(bytes)}\n`);
  return 0;
}

function main(argv: string[]): number {
  const ver = requireNodeVersion();
  if (ver !== 0) return ver;
  const root = repoRootFromGateFile(import.meta.url);
  const args = argv.slice(2);
  if (
    args.length === 0 ||
    args[0] === "-h" ||
    args[0] === "--help" ||
    args[0] === "help"
  ) {
    writeOut("usage: node tools/gate/gate.ts status|hash <file>\n");
    writeOut("W1 stub: status + hash (DEC-144). Other subcommands land in W2.\n");
    return 0;
  }
  const cmd = args[0] ?? "";
  const rest = args.slice(1);
  if (cmd === "status") return cmdStatus(root);
  if (cmd === "hash") return cmdHash(rest);
  writeErr(`unknown command ${cmd}; W1 stub supports: status, hash\n`);
  return 2;
}

process.exitCode = main(process.argv);

File: tools/gate/gate.py
#!/usr/bin/env python3
# keel gate — W1 stub. Real subcommands (check/new/index/trace/verify/…) land in W2.
# Python ≥3.11, stdlib only (C-101). Windows: python -X utf8.

from __future__ import annotations

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[2]
RECORDS = ROOT / "keel"
CONFIG = RECORDS / "config.json"


def load_config() -> dict:
    if not CONFIG.is_file():
        return {}
    data = json.loads(CONFIG.read_text(encoding="utf-8"))
    return {k: v for k, v in data.items() if not str(k).startswith("_")}


def cmd_status(argv: list[str]) -> int:
    cfg = load_config()
    tier = cfg.get("enforcement_tier", "unknown")
    print("keel status")
    print("wave: W1-stub")
    print(f"root: {ROOT}")
    print(f"records_dir: {RECORDS}")
    print(f"handoff: {RECORDS / 'handoff.md'}")
    print(f"overview: {RECORDS / 'OVERVIEW.md'}")
    print(f"plan_index: {RECORDS / 'plan' / 'INDEX.md'}")
    print(f"enforcement_tier: {tier}")
    print("note: full gates are W2; this stub only prints start-of-session paths (C-27/C-72).")
    if argv:
        print("ignored extra args: " + " ".join(argv), file=sys.stderr)
    return 0


def main(argv: list[str] | None = None) -> int:
    args = list(sys.argv[1:] if argv is None else argv)
    if not args or args[0] in {"-h", "--help", "help"}:
        print("usage: python -X utf8 tools/gate/gate.py status")
        print("W1 stub: only `status` is implemented.")
        return 0
    cmd, *rest = args
    if cmd == "status":
        return cmd_status(rest)
    print(f"unknown command {cmd!r}; W1 stub supports: status", file=sys.stderr)
    return 2


if __name__ == "__main__":
    sys.exit(main())
